src/java.naming/share/classes/com/sun/jndi/ldap/AbstractLdapNamingEnumeration.java

```diff
@@ -195,49 +285,62 @@ public final boolean hasMore() throws NamingException {
+            if (!hasMoreCalled) {
+                hasMore();
+            }
+            hasMoreCalled = false;
+            return nextImpl();
+        } finally {
+            // Ensure writes are visible to the Cleaner thread
+            VarHandle.fullFence();
+            // Ensure Cleaner does not run until after this method completes
+            Reference.reachabilityFence(enumCtx);
         }
-        hasMoreCalled = false;
-        return nextImpl();
     }
 
     /*
      * Test if unprocessed entries or referrals exist.
      */
     private boolean hasMoreImpl() throws NamingException {
-        // when page size is supported, this
-        // might generate an exception while attempting
-        // to fetch the next batch to determine
-        // whether there are any more elements
-
-        // test if the current set of entries has been processed
-        if (posn == limit) {
-            getNextBatch();
-        }
-
-        // test if any unprocessed entries exist
-        if (posn < limit) {
-            return true;
-        } else {
-
-            try {
-                // try to process another referral
-                return hasMoreReferrals();
+        try {
+            // when page size is supported, this
+            // might generate an exception while attempting
+            // to fetch the next batch to determine
+            // whether there are any more elements
+
+            // test if the current set of entries has been processed
+            if (posn == limit) {
+                getNextBatch();
+            }
 
-            } catch (LdapReferralException |
-                     LimitExceededException |
-                     PartialResultException e) {
-                cleanup();
-                throw e;
+            // test if any unprocessed entries exist
+            if (posn < limit) {
+                return true;
+            } else {
 
-            } catch (NamingException e) {
-                cleanup();
-                PartialResultException pre = new PartialResultException();
-                pre.setRootCause(e);
-                throw pre;
+                try {
+                    // try to process another referral
+                    return hasMoreReferrals();
+
+                } catch (LdapReferralException |
+                        LimitExceededException |
+                        PartialResultException e) {
+                    cleanup();
+                    throw e;
+
+                } catch (NamingException e) {
+                    cleanup();
+                    PartialResultException pre = new PartialResultException();
+                    pre.setRootCause(e);
+                    throw pre;
+                }
             }
+        } finally {
+            // Ensure writes are visible to the Cleaner thread
+            VarHandle.fullFence();
+            // Ensure Cleaner does not run until after this method completes
+            Reference.reachabilityFence(enumCtx);
```

Comment:
The same as above: no need for memory fence and reachability fence should take 'this' as parameter.